lib/src/Base/Optim/LeastSquaresProblem.cxx

```diff
@@ -0,0 +1,110 @@
+//                                               -*- C++ -*-
+/**
+ *  @brief LeastSquaresProblem allows to describe an optimization problem
+ *
+ *  Copyright 2005-2019 Airbus-EDF-IMACS-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/LeastSquaresProblem.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/QuadraticFunction.hxx"
+#include "openturns/ComposedFunction.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(LeastSquaresProblem)
+
+static const Factory<LeastSquaresProblem> Factory_LeastSquaresProblem;
+
+/* Default constructor */
+LeastSquaresProblem::LeastSquaresProblem()
+  : OptimizationProblemImplementation()
+{
+  // Nothing to do
+}
+
+LeastSquaresProblem::LeastSquaresProblem(const Function & residualFunction)
+  : OptimizationProblemImplementation()
+{
+  setResidualFunction(residualFunction);
+}
+
+/* Virtual constructor */
+LeastSquaresProblem * LeastSquaresProblem::clone() const
+{
+  return new LeastSquaresProblem(*this);
+}
+
+/* Objective accessor */
+Function LeastSquaresProblem::getResidualFunction() const
+{
+  return residualFunction_;
+}
+
+void LeastSquaresProblem::setResidualFunction(const Function & residualFunction)
+{
+  // F(p)_i = f(x_i, p)-y_i
+  // g(p)= sum_i 0.5*(F(p)_i)^2
+  const UnsignedInteger residualSize = residualFunction.getOutputDimension();
+//   Description resDesc(Description::BuildDefault(residualSize, "R"));
+//   String formula("0");
+//   for (UnsignedInteger i = 0; i < residualSize; ++ i)
+//     formula += "+0.5*" + resDesc[i] + "^2";
+//   SymbolicFunction sumQuad(resDesc, Description(1, formula));
+
+  residualFunction_ = residualFunction;
+  const Point center(residualSize);
+  const Point constant(1);
+  const Matrix linear(residualSize, 1);
+  SymmetricTensor quadratic(residualSize, 1);
+  const QuadraticFunction sumQuad(center, constant, linear, quadratic);
+  for (UnsignedInteger i = 0; i < residualSize; ++ i)
+    quadratic(i, i, 0) = 0.5;
+  objective_ = ComposedFunction(sumQuad, residualFunction);
```

Comment:
Not the most efficient way to compute half of the squared norm of a vector, but at least it exists and it works ;-)! I will add a dedicated function to compute u->||u||^2/2, its gradient u->u and its Hessian u->Id later on.